src/CodeStyle/Core/Analyzers/Formatting/Formatter.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Formatting.Rules;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Formatting
+{
+    /// <summary>
+    /// Formats whitespace in documents or syntax trees.
+    /// </summary>
+    internal static class Formatter
```

Comment:
📝 Copied/modified/pruned from workspaces layer. It would be difficult to share this file due to signature changes related to dependency injection strategies.

Reply:
Can you expand on `signature changes related to dependency injection strategies`? You seem to have most types used in this file available in both the projects? Not blocking on sharing this code, just curious to know what were the difficulties..

---
In reply to: [234982688](https://github.com/dotnet/roslyn/pull/31276#discussion_r234982688) [](ancestors = 234982688)

Reply:
Just looked at [http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Formatting/Formatter.cs,38](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Formatting/Formatter.cs,38) and figured out most APIs there are based off `document.GetLanguageService<IFormattingService>();` or `document.GetLanguageService<ISyntaxFormattingService>()`, so this seems to make sense now.

---
In reply to: [236819754](https://github.com/dotnet/roslyn/pull/31276#discussion_r236819754) [](ancestors = 236819754,234982688)